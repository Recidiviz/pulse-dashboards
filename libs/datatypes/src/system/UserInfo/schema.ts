// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { z } from "zod";

import { supervisionOfficerSupervisorSchema } from "~datatypes";

export const userInfoSchema = z
  .discriminatedUnion("role", [
    z.object({
      role: z.literal("supervision_officer_supervisor"),
      entity: supervisionOfficerSupervisorSchema,
    }),
    z.object({ role: z.null(), entity: z.null() }),
  ])
  .and(
    // TODO(#4651): Remove union type once backend sends metadata
    z.union([
      z.object({
        hasSeenOnboarding: z.boolean(),
        metadata: z.undefined(),
      }),
      z.object({
        hasSeenOnboarding: z.undefined(),
        metadata: z.object({
          hasSeenOnboarding: z.boolean(),
        }),
      }),
    ]),
  )
  .transform(({ hasSeenOnboarding, metadata, ...rest }) => ({
    metadata: {
      hasSeenOnboarding:
        metadata?.hasSeenOnboarding ?? hasSeenOnboarding ?? false, // this "false" should never occur due to the definitions above, but typescript is only _so_ smart, so provide it as a fallback
    },
    ...rest,
  }));

export type UserInfo = z.infer<typeof userInfoSchema>;
export type RawUserInfo = z.input<typeof userInfoSchema>;
