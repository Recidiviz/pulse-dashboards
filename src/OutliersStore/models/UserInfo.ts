// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { z } from "zod";

import { supervisionOfficerSupervisorSchema } from "./SupervisionOfficerSupervisor";

export const userInfoSchema = z.discriminatedUnion("role", [
  z.object({
    role: z.literal("supervision_officer_supervisor"),
    entity: supervisionOfficerSupervisorSchema,
  }),
  z.object({ role: z.null(), entity: z.null() }),
]);

export type UserInfo = z.infer<typeof userInfoSchema>;
export type RawUserInfo = z.input<typeof userInfoSchema>;
