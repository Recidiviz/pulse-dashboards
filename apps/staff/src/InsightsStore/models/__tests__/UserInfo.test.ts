// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  rawLeadershipUserInfoFixture,
  rawSupervisorUserInfoFixture,
} from "../offlineFixtures/UserInfoFixture";
import { RawUserInfo, userInfoSchema } from "../UserInfo";

test("supervisor transformation", () => {
  expect(userInfoSchema.parse(rawSupervisorUserInfoFixture)).toMatchSnapshot();
});

test("leadership transformation", () => {
  expect(userInfoSchema.parse(rawLeadershipUserInfoFixture)).toMatchSnapshot();
});

test("null entity with supervisor role fails", () => {
  expect(() => {
    userInfoSchema.parse({
      ...rawSupervisorUserInfoFixture,
      entity: null,
    });
  }).toThrow();
});

test("present entity with null role fails", () => {
  expect(() => {
    userInfoSchema.parse({
      ...rawSupervisorUserInfoFixture,
      role: null,
    });
  }).toThrow();
});

test("hasSeenOnboarding is moved into metadata", () => {
  const testUserInfo: RawUserInfo = {
    entity: null,
    role: null,
    hasSeenOnboarding: true,
  };
  expect(userInfoSchema.parse(testUserInfo)).toMatchInlineSnapshot(`
    {
      "entity": null,
      "metadata": {
        "hasSeenOnboarding": true,
      },
      "role": null,
    }
  `);
});
