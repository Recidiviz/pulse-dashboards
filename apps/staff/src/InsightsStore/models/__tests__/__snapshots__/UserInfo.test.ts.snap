// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`leadership transformation 1`] = `
{
  "entity": null,
  "metadata": {
    "hasSeenOnboarding": false,
  },
  "role": null,
}
`;

exports[`supervisor transformation 1`] = `
{
  "entity": {
    "displayName": "Miles D Davis",
    "email": "mock-email",
    "externalId": "mdavis123",
    "fullName": {
      "givenNames": "Miles",
      "middleNames": "D",
      "surname": "Davis",
    },
    "hasOutliers": true,
    "pseudonymizedId": "hashed-mdavis123",
    "supervisionDistrict": "Region D1",
  },
  "metadata": {
    "hasSeenOnboarding": false,
  },
  "role": "supervision_officer_supervisor",
}
`;
